### PR TITLE
Adding ObjectSelector to admission webhooks

### PR DESCRIPTION
**Pull request by caesarxuchao**
Supersedes #78226.

Based on https://github.com/kubernetes/kubernetes/pull/78491 and https://github.com/kubernetes/kubernetes/pull/78080, which are both reviewed and approved already

Implementing [KEP#objectSelector](https://github.com/liggitt/enhancements/blob/7b7fc0c2293c02c9bb580805dbdae49d802a189b/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md#object-selector) and https://github.com/kubernetes/kubernetes/pull/78226#issuecomment-497102966.

/sig api-machinery
/assign @liggitt @jpbetz @sttts @roycaihw 

```release-note
Added objectSelector to admission webhook configurations. objectSelector is evaluated the oldObject and newObject that would be sent to the webhook, and is considered to match if either object matches the selector. A null object (oldObject in the case of create, or newObject in the case of delete) or an object that cannot have labels (like a DeploymentRollback or a PodProxyOptions object) is not considered to match. Use the object selector only if the webhook is opt-in, because end users may skip the admission webhook by setting the labels.
````

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/object/matcher.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+// Matcher decides if a request selected by the ObjectSelector.
+type Matcher struct {
+}
+
+func matchObject(obj runtime.Object, selector labels.Selector) bool {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		klog.V(4).Infof("cannot access metadata of %v: %v", obj, err)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/object/matcher.go]: given how frequently this will be logged, probably V(5)

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/object/matcher.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/klog"
+)
+
+// Matcher decides if a request selected by the ObjectSelector.
+type Matcher struct {
+}
+
+func matchObject(obj runtime.Object, selector labels.Selector) bool {
+	accessor, err := meta.Accessor(obj)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/object/matcher.go]: check if obj is nil and return false first?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/object/matcher.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+		return false
+	}
+	return selector.Matches(labels.Set(accessor.GetLabels()))
+
+}
+
+// MatchObjectSelector decideds whether the request matches the ObjectSelector
+// of the webhook. Only when they match, the webhook is called.
+func (m *Matcher) MatchObjectSelector(h *v1beta1.Webhook, attr admission.Attributes) (bool, *apierrors.StatusError) {
+	selector, err := metav1.LabelSelectorAsSelector(h.ObjectSelector)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/object/matcher.go]: add a TODO to cache this selector. once this gets rebased on https://github.com/kubernetes/kubernetes/pull/78491, we'll likely want to make GetLabelSelector() part of the webhook accessor and construct it once when we construct the accessor

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go [hunk 4]
```diff
@@ -183,6 +186,11 @@ func (a *Webhook) shouldCallHook(h *v1beta1.Webhook, attr admission.Attributes,
 		return nil, err
 	}
 
+	matches, err = a.objectMatcher.MatchObjectSelector(h, attr)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go]: shouldCallHook is currently called on all hooks before dispatching to any of them. we wanted to call this inside the loop in mutatingDispatcher#Dispatch so the labels we evaluate on are the ones on the objects sent to the webhook

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go]: that may mean Dispatch needs to be given an unfiltered list of webhooks and a way to invoke the `shouldCallHook` function

File: staging/src/k8s.io/api/admissionregistration/v1beta1/types.go [hunk 5]
```diff
@@ -252,13 +252,27 @@ type Webhook struct {
...
+	// is considered to match if either object matches the selector. A null
+	// object (oldObject in the case of create, or newObject in the case of
+	// delete) or an object that cannot have labels (like a
+	// DeploymentRollback or a PodProxyOptions object) is not considered to
+	// match.
+	// Use the object selector only if the webhook is opt-in, because end
+	// users may skip the admission webhook by setting the labels.
+	// Default to the empty LabelSelector, which matches everything.
+	// +optional
+	ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty" protobuf:"bytes,10,opt,name=objectSelector"`
```
Comment (liggitt) [inline staging/src/k8s.io/api/admissionregistration/v1beta1/types.go]: this needs to be validated to be sure we can successfully construct a label selector from it

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/object/matcher.go (hunk 2 above)
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/object/matcher.go]: meta.Accessor will return an err, which causes `false` to be returned.

Or is your comment more about readability?

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/object/matcher.go]: isn't the nil check cheaper?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go [hunk 6]
```diff
@@ -39,28 +39,45 @@ import (
...
 		// If we already have this version, continue
-		if _, ok := versionedAttrs[call.Kind]; ok {
+		if _, ok := versionedAttrs[invocation.Kind]; ok {
 			continue
 		}
-		versionedAttr, err := generic.NewVersionedAttributes(attr, call.Kind, o)
+		versionedAttr, err2 := generic.NewVersionedAttributes(attr, invocation.Kind, o)
 		if err != nil {
-			return apierrors.NewInternalError(err)
+			return apierrors.NewInternalError(err2)
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go]: `if err2 != nil` ? Or can we just use `err` here as well?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/object/matcher.go (hunk 2 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/object/matcher.go]: I'd like it just for readability. It makes it clear that we are being deliberate in how nil is handled, and that it doesn't result in false by accident.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/interfaces.go [hunk 7]
```diff
@@ -60,6 +60,6 @@ type WebhookInvocation struct {
 
 // Dispatcher dispatches webhook call to a list of webhooks with admission attributes as argument.
 type Dispatcher interface {
-	// Dispatch a request to the webhooks using the given webhooks. A non-nil error means the request is rejected.
-	Dispatch(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces, hooks []*WebhookInvocation) error
+	// Dispatch a request to the webhooks. Dispatcher may choose not to call a hook. A non-nil error means the request is rejected.
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/interfaces.go]: this needs to be crisper about specifically why the webhook can choose not to call the hook

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/interfaces.go [hunk 8]
```diff
@@ -60,6 +60,6 @@ type WebhookInvocation struct {
...
+	Dispatch(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces, hooks []v1beta1.Webhook) error
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/interfaces.go]: this will need rebasing on https://github.com/kubernetes/kubernetes/pull/78080

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go [hunk 9]
```diff
@@ -55,10 +55,16 @@ func newMutatingDispatcher(p *Plugin) func(cm *webhook.ClientManager) generic.Di
 
 var _ generic.Dispatcher = &mutatingDispatcher{}
 
-func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces, relevantHooks []*generic.WebhookInvocation) error {
+func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces, hooks []v1beta1.Webhook) error {
 	var versionedAttr *generic.VersionedAttributes
-	for _, invocation := range relevantHooks {
-		hook := invocation.Webhook
+	for _, hook := range hooks {
+		invocation, statusErr := a.plugin.ShouldCallHook(&hook, attr, o)
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: we need an integration test demonstrating that a label change by webhook1 affects selection by webhook2. copying Joe's `test/integration/apiserver/admissionwebhook/reinvocation_test.go` as a starting point and adding object selection would probably be quickest... it lets you easily combine different orders of webhooks and verify they get called (or don't get called) as expected.

as written, o.GetObject() only gets updated if webhooks need different versions of an object (which forces conversion via o.GetObject()) or after we exit this loop. that means if webhook1 modified labels, the selector could be checking stale labels for webhook2

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go (hunk 6 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go]: we can just use err here as well

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (caesarxuchao) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: Added unit test cases to demonstrate the behavior.

Fixed the problem Jordan raised.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/interfaces.go [hunk 10]
```diff
@@ -35,7 +35,7 @@ type Source interface {
 // variants of the object and old object.
 type VersionedAttributes struct {
 	// Attributes holds the original admission attributes
-	Attributes admission.Attributes
+	admission.Attributes
```
Comment (roycaihw) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/interfaces.go]: is this change intended?

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/interfaces.go]: it is intended... it allows passing versionedattributes to ShouldCallHook as admission.Attributes and overriding the GetObject() method

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/interfaces.go [hunk 11]
```diff
@@ -48,6 +48,14 @@ type VersionedAttributes struct {
 	Dirty bool
 }
 
+// GetObject overrides the Attributes.GetObject()
+func (v *VersionedAttributes) GetObject() runtime.Object {
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/interfaces.go]: this is fine for now, and we have complete test coverage, but a follow-up issue to rework this a little in 1.16 would be good

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go [hunk 12]
```diff
@@ -38,28 +39,45 @@ import (
...
 
 type validatingDispatcher struct {
-	cm *webhookutil.ClientManager
+	cm     *webhookutil.ClientManager
+	plugin *Plugin
 }
 
-func newValidatingDispatcher(cm *webhookutil.ClientManager) generic.Dispatcher {
-	return &validatingDispatcher{cm}
+func newValidatingDispatcher(p *Plugin) func(cm *webhookutil.ClientManager) generic.Dispatcher {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go]: this works for now, but is pretty awkward, can you open a follow-up issue to avoid needing to inject the full plugin into the dispatchers?

**Comment by liggitt** [comment]
looks like a compile error in `k8s.io/apiserver/pkg/admission/plugin/webhook/object`, and openapi spec needs updating

**Comment by liggitt** [comment]
ran against https://github.com/kubernetes/kubernetes/pull/78309, v1.14.0 compatibility looked good, new fields showed up as expected in the data fixtures for HEAD

**Comment by liggitt** [comment]
/lgtm
/approve

API change looks good

**Comment by liggitt** [comment]
included in docs PR for 1.15 at https://github.com/kubernetes/website/pull/14671

**Comment by liggitt** [comment]
> This looks like a breaking change.

The type name changed, but the wire serialization of the two types is identical.
